openedx/openedx-learning
# Single taxonomy view [FC-0030]

## Description

Adds a new view to return tags for a given taxonomy. With this view, you can:

- Return paginated root tags.
- Return paginated children of given tag.
- Search tags in all taxonomy.
- Search tags on a children list.

Also updates the ADR 0014 with some logics on search.

## Supporting Information

- Closes https://github.com/openedx/modular-learning/issues/75

## Testing Instructions

- Ensure that the tests cover the expected behavior of the view as described in the related issue.
- Run this command to create a small taxonomy:

```
python manage.py loaddata tests/openedx_tagging/core/fixtures/tagging.yaml
```

- Run this commands to create a large taxonomy:

```
python manage.py shell
> from tests.openedx_tagging.core.tagging.test_views import TestTaxonomyTagsView
> case = TestTaxonomyTagsView()
> case.setUp()
> case.build_large_taxonomy()
> from openedx_tagging.core.tagging.models import Taxonomy, Tag
> Taxonomy.objects.filter(name="Large Taxonomy")
```

- Run django server: `python manage.py runserver`
- You can comment [this line](https://github.com/openedx/openedx-learning/pull/78/files#diff-215d20c663dcd5727c14740a72d09fe6fd22e3a92fd08db46036e2e799699917R354) to avoid the permissions.
- Make a [Get request](http://127.0.0.1:8000/tagging/rest_api/v1/taxonomies/-1/tags/) to get the Languages tags. Verify that all tags are returned. Verify the structure.
- Make a Get request to get the small taxonomy tags: http://127.0.0.1:8000/tagging/rest_api/v1/taxonomies/Tax-id/tags/
- Verify that all tags are returned. Verify the tree structure.
- Make a Get request to get the large taxonomy tags.
    - Verify that the root tags are paginated.
    - Verify the structure.
    - Open the next page link.
    - Verify the result.
    - Open a children link.
    - Verify the result.
- Make a Get request to search tags on the small taxonomy:  http://127.0.0.1:8000/tagging/rest_api/v1/taxonomies/Tax-id/tags/?search_term=eU
- Verify the results.
- Make a Get request to search tags on the large taxonomy: http://127.0.0.1:8000/tagging/rest_api/v1/taxonomies/Tax-id/tags/?search_term=10
- Verify the results.
